### PR TITLE
Ensure sending stream data also includes outstanding ACKs.

### LINKED ISSUE #865: Sending application data does not override ACK delay
This means that when an application sends data, it's not sending any pending ACKs, which are waiting for the ACK delay timer to expire before being sent.
This was discovered via interop client and could be affecting pass rates.

### PR DESCRIPTION
**Pull request by anrossi**
Fixes #865.

### CODE REVIEW

File: src/core/connection.h [hunk 1]
```diff
@@ -254,6 +254,8 @@ typedef struct QUIC_CONN_STATS {
 
         uint64_t TotalBytes;            // Sum of UDP payloads
         uint64_t TotalStreamBytes;      // Sum of stream payloads
+
+        uint64_t AcksReceived;          // Count of receive ACK frames.
```
Comment (nibanks) [inline src/core/connection.h:258]: Don't put Received in the name. Maybe, call it `ValidAckFrames` and put it after `ValidPackets`.

File: src/test/MsQuicTests.h [hunk 2]
```diff
@@ -235,6 +235,10 @@ QuicTestReceiveResumeNoData(
     _In_ QUIC_RECEIVE_RESUME_SHUTDOWN_TYPE ShutdownType
     );
 
+void
+QuicTestAckSendDelay(
+);
```
Comment (nibanks) [inline src/test/MsQuicTests.h:240]: nit: tab over
```suggestion
    );
```

File: src/test/lib/DataTest.cpp [hunk 3]
```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
...
+                nullptr);
+            if (QUIC_FAILED(Status)) {
+                TEST_FAILURE("Server failed to send to send data back 0x%x", Status);
+            }
+            break;
+        default:
+            break;
+        }
+    } else {
+        if(TestContext->ClientStream.Handle != QuicStream) {
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:1912]:
```suggestion
        if (TestContext->ClientStream.Handle != QuicStream) {
```

File: src/test/lib/DataTest.cpp [hunk 4]
```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
...
+        case QUIC_STREAM_EVENT_RECEIVE:
+            Event->RECEIVE.TotalBufferLength = 0;
+            Status = MsQuic->StreamSend(
+                QuicStream,
+                TestContext->SendBuffer.Buffers,
+                TestContext->SendBuffer.BufferCount,
+                QUIC_SEND_FLAG_FIN,
+                nullptr);
+            if (QUIC_FAILED(Status)) {
+                TEST_FAILURE("Server failed to send to send data back 0x%x", Status);
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:1905]: `TEST_FAILURE` doesn't return. You should put explicit return calls after each of them in this function.

File: src/test/lib/DataTest.cpp [hunk 5]
```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
...
+    uint32_t AckDelayMs = 1000;
+    MsQuicRegistration Registration;
+    TEST_TRUE(Registration.IsValid());
+
+    MsQuicAlpn Alpn("MsQuicTest");
+
+    MsQuicSettings Settings{};
+    Settings.SetIdleTimeoutMs(TimeoutMs);
+    Settings.MaxAckDelayMs = AckDelayMs;
+    Settings.IsSet.MaxAckDelayMs = true;
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:2033]: I feel like we should have a rule: If you add code to set a setting that doesn't have a helper yet, you get to add the helper. :D

File: src/test/lib/DataTest.cpp [hunk 6]
```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
...
+    Settings.SetPeerBidiStreamCount(1);
+
+    MsQuicConfiguration ServerConfiguration(Registration, Alpn, Settings, SelfSignedCredConfig);
+    TEST_TRUE(ServerConfiguration.IsValid());
+
+    MsQuicCredentialConfig ClientCredConfig;
+    MsQuicConfiguration ClientConfiguration(Registration, Alpn, Settings, ClientCredConfig);
+    TEST_TRUE(ClientConfiguration.IsValid());
+
+    QUIC_ADDRESS_FAMILY QuicAddrFamily = (true) ? QUIC_ADDRESS_FAMILY_INET : QUIC_ADDRESS_FAMILY_INET6;
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:2043]: A TODO to add family as input to this test?

File: src/test/lib/DataTest.cpp (hunk 5 above)
Comment (anrossi) [reply src/test/lib/DataTest.cpp:2040]: Good rule!

**Review by nibanks — APPROVED** [review]

**Review by ThadHouse — APPROVED** [review]

**Pull request merged**